mlir/include/mlir/IR/Attributes.h

```diff
@@ -322,12 +322,19 @@ class AttributeInterface
 // Core AttributeTrait
 //===----------------------------------------------------------------------===//
 
-/// This trait is used to determine if an attribute is mutable or not. It is
-/// attached on an attribute if the corresponding ImplType defines a `mutate`
-/// function with proper signature.
 namespace AttributeTrait {
+/// This trait is used to determine if an attribute is mutable or not. It is
+/// attached on an attribute if the corresponding ConcreteType defines a
+/// `mutate` function with proper signature.
 template <typename ConcreteType>
 using IsMutable = detail::StorageUserTrait::IsMutable<ConcreteType>;
+
+/// This trait is used to determine if an attribute is a location or not. It is
```

Comment:
What's the trade off of having a trait for this, vs having an interface? My original expectation for this kind of patch was to effectively just turn LocationAttr into an interface. I suppose going to an interface would be much more costly, but what happens to the API on LocationAttr (e.g. walk?). How do dialect interfaces hook into that handling?

Reply:
I think the trade-off is what you just mentioned - it's a little less flexible, but it's also less costly. My major concern was that when parsing, we don't want every location parsed to have to go through interface dispatch, basically.

> but what happens to the API on LocationAttr (e.g. walk?)

Wouldn't the walk just advance into the attribute itself with `walkFn(*this)` up at the top? I could totally be missing something but that was my read. The thought was that if you called `walk` you'd call the generic attribute `walk`, which does the same walk you'd expect from anywhere.

Reply:
Discussed offline - I misunderstood the purpose of `walk` - I have an update that I'll push shortly that should allow you to maintain the current behavior of only walking sub-locations.

Reply:
Updated and added a test